### ISSUE TITLE
Spurious unbound variables when declaring variables in an "if" block

### ISSUE THREAD

**Issue by FedericoV** [microsoft/pyright#219]
**Describe the bug**
If a variable is only declared within a conditional scope, and then, used later down the in line in a similar conditional scope, pyright warns about out of bound concerns.

**To Reproduce**
```
   if use_dask:
        dd_test_data_df = dd.from_pandas(
            target_df[cat_columns], npartitions=n_partitions
        )

    # I did some experiments, and row-based parallelism with dask ends up being much faster than
    # executing the encoding in parallel across columns with joblib.
    for col_name in cat_columns:
        col_encoding = categorical_encoding[col_name]

        # Annoyingly, we cannot use replace (which would be much faster) because it does not like
        # dictionaries with loops.  Very annoying.
        if use_dask:
            meta = pd.Series(index=target_df.index, name=col_name, dtype=np.int32)
            result = dd_test_data_df[col_name].map(col_encoding, meta=meta)
            target_df[col_name] = result.compute()
        else:
            target_df[col_name] = target_df[col_name].map(col_encoding)
```

pyright warns that this line:
```
result = dd_test_data_df[col_name].map(col_encoding, meta=meta)
```
is possibly unbound.
**Expected behavior**

Ideally, pyright should catch that the variable is only used under the same situations where it is instantiated.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```
def foo(self) -> str:
    return 3
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

**Additional context**

**Comment by erictraut** [comment]
Tracking whether the conditionals is not possible in the general case because some expressions have side effects or rely on state that can change between evaluations. There are limited cases where this could be determined statically, but I don't think it's worth special-casing these. Instead, I recommend avoiding this pattern. One easy workaround is to pre-initialize the variable (e.g. with None). Another easy workaround in your case is to change the second conditional from `if use_dask` to `if dd_test_data_df`.

**Comment by sid-khullar** [comment]
FYI: Pre-declaring variables doesn't remove the notification in all cases.

**Comment by taldcroft** [comment]
I am having the same issue, and this supports the request in #144 to have some mechanism for disabling a warning on a specific line when it is not possible for a static checker to track the conditionals.

**Comment by taldcroft** [comment]
Do you have a suggestion to avoid a spurious warning for this pattern:
```
try:
    for name in names:
        do_something(name)
except Exception as err:
    # Do some handling for this case
    print(f'do_something({name}) failed: {err}')
else:
    post_do_something_processing()
```

**Comment by erictraut** [comment]
My recommendation is to assign `None` to `name` before the `for` loop. If the iterator `names` generates an exception, `name` will be unbound, and it will generate another exception within the `except` handler.

**Comment by taldcroft** [comment]
> My recommendation is to assign None to name before the for loop

That solves the problem, but at the expense of introducing code that has no purpose whatsoever except to satisfy pyright. In this case (and rather commonly), `names` is a Python `list` and the language guarantees that iterating over that list will not generate an exception. 

In going through my code I have already had to add useless code in several instances (for different reasons) in order to keep pyright happy. The other Python code checkers I've used all accept that there are valid reasons for a developer to skip particular checks. OTOH it did find some places that needed improvement, so I am very grateful that this package is being developed.

Is there a technical reason that it is not possible to disable pyright checks on a particular code line?
